text/0011-html_viz.md

```diff
@@ -0,0 +1,182 @@
+| default     | allows setting a default value that can be returned if a first value is not set. |
+| title       | give the title of a dataset |
+| getBody     | load the full dataset body |
+| filesize    | convert byte count to kb/mb/etc string |
+
+
+#### future dataset document API
+
+We have reserved future work for a "dataset API" that will expand the default capabilities of a dataset document to include convenience functions for doing things like loading named columns or sampling rows from the body. We've intentionally left this API undefined thus far to understand how it will work in different contexts. One such context is this template API.
+
+The one exception to this is exposing body data through a global function: `getBody`. This is because there's a _very_ high chance we'll want to export `ds.body` as an object with methods in the future. If we simply load the entire body & drop it into `ds.body`, adding methods to `ds.body` will require breaking the document API.
+
+### The default template
+
+Our standard template should be a collection of pre-defined blocks which are also available to user-provided templates. An example default template would look something like this:
+
+```html
+<!DOCTYPE html>
+<html>
+  <head>
+  {{ block "stylesheet" . }}{{ end }}
+  </head>
+  <body>
+  {{ block "header" ds . }}{{ end }}
+  {{ block "summary" ds . }}{{ end }}
+  {{ block "stats" ds . }}{{ end }}
+  {{ block "citations" ds . }}{{ end }}
+  </body>
+</html>
+```
+
+Users can then swap in these predefined blocks to make pseudo-custom templates a thing, and ease the transition to fully-custom visualizations through progressive customization. The pre-defined blocks are as follows:
+
+| name       | purpose |
+| ---------- | ------- |
+| stylesheet | default css styles |
+| header     | dataset title, date created & author in a `<header>` tag |
+| summary    | overview of dataset details |
+| stats      | template of stats component that prints nothing if the stats component is undefined |
+| citations  | `<footer>` tag that prints hash content & links to in-package files |
+
+In practice I'm hoping the `citations` partial will get a lot of use, making it easier to drop properly-formatted citations into markup with little thought.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+One of the biggest drawbacks to the proposed implementation comes from a natural tendency to use the following to template the contents of a dataset body in to a dataset:
+
+```html
+<script type="text/javascript"> window.data = {{ ds.get_body }} </script>
+```
+
+Using templating this way prints the entire body into the template. If the body is large, this will cause _all body data to be duplicated in `index.html`_. I propose we fight this through documentation, because there are many scenarios where this code is completely appropriate (for example, when the dataset body is an aggregation of some other data source).
```

Comment:
I'm not sure we should even support this scenario. If someone wants to access the entirety of the body, perhaps we should force them to explicitly copy the body into the stats somewhere. That is, by requiring additional, explicit steps, instead of providing access by default, we would imply that it's not a recommended course of action.

Reply:
Yeah, my concern is there are situations where it absolutely _is_ the recommended course of action. For instance, [b5/country_codes](https://app.qri.io/b5/country_codes) only has 250 entries, If for some reason we needed to graph those, It would be fine to embed the entire body.